multiversx_sdk_core/transaction_factories/smart_contract_factory.py

```diff
@@ -0,0 +1,192 @@
+from pathlib import Path
+from typing import Any, List, Optional, Protocol, Tuple
+
+from Cryptodome.Hash import keccak
+
+from multiversx_sdk_core.address import Address
+from multiversx_sdk_core.constants import (ARGS_SEPARATOR,
+                                           CONTRACT_DEPLOY_ADDRESS,
+                                           DEFAULT_HRP,
+                                           TRANSACTION_OPTIONS_DEFAULT,
+                                           TRANSACTION_VERSION_DEFAULT,
+                                           VM_TYPE_WASM_VM)
+from multiversx_sdk_core.interfaces import (IAddress, IChainID, IGasLimit,
+                                            IGasPrice, INonce,
+                                            ITransactionValue)
+from multiversx_sdk_core.serializer import arg_to_string
+from multiversx_sdk_core.transaction import Transaction
+from multiversx_sdk_core.transaction_payload import TransactionPayload
+from multiversx_sdk_core.utils import read_binary_file
+
+
+class IConfig(Protocol):
+    chain_id: IChainID
+
+
+class CodeMetadata:
+    def __init__(self, upgradeable: bool = True, readable: bool = True, payable: bool = False, payable_by_sc: bool = False) -> None:
+        self.upgradeable = upgradeable
+        self.readable = readable
+        self.payable = payable
+        self.payable_by_sc = payable_by_sc
+
+    def to_hex(self) -> str:
+        flag_value_pairs = [
+            (0x01_00, self.upgradeable),
+            (0x04_00, self.readable),
+            (0x00_02, self.payable),
+            (0x00_04, self.payable_by_sc)
+        ]
+        metadata_value = self.sum_flag_values(flag_value_pairs)
+        return f"{metadata_value:04X}"
+
+    def sum_flag_values(self, flag_value_pairs: List[Tuple[int, bool]]) -> int:
```

Comment:
Even though this addition works for bit flags, in other implementations we are explicitly using `bitwise or`.

E.g.:
 - https://github.com/multiversx/mx-chain-vm-common-go/blob/feat/freeze-account/codeMetadata.go
 - https://github.com/multiversx/mx-sdk-js-core/blob/main/src/smartcontracts/codeMetadata.ts

Do you think we can have an implementation that is best mirroring the Go one (the "protocol one")? Also, it can be extracted to a separate file.